### PR TITLE
documentation for .yeb format

### PR DESCRIPTION
**Pull request by Caylo**
Compiled version: http://caylo-easybuild.readthedocs.org/en/yeb-docs/Writing_yeb_easyconfig_files.html

### CODE REVIEW

**Comment by rjeschmi** [comment]
Did we get the docs for the yeb options anywhere yet?

**Comment by boegel** [comment]
What do you mean with yeb options?

**Comment by rjeschmi** [comment]
Maybe what I'm looking for is pre-release docs. We could put them somewhere else (ie, checkout feature branch, use --experimental etc).

Also, the example in the docs seems to be slightly wrong (easyblock shouldn't be specified, or should be bzip I think, since there is no configure script used for the sources)

**Comment by rjeschmi** [comment]
Also the escaping in the docs for *SOURCE_TAR_GZ (shows up as \*SOURCE_TAR_GZ).

Is * a preference or something that makes sense to YAML? It is going to be a bit of a pain long term I imagine (for the same reason the docs are hard to escape).

**Comment by boegel** [comment]
We currently don't have documentation for pre-release stuff, but we can add instructions here on how to start playing around with this.
We can even provide an easyconfig for to install EasyBuild 2.x.0dev with .yeb support included.

**Comment by rjeschmi** [comment]
that might be nice. This is a really cool feature. Definitely looking forward to this moving forward.

**Comment by boegel** [comment]
And you're right about the `easyblock` thing, we took a shortcut there to use a `bzip2` as a test case, to avoid requiring the `EB_bzip2` easyblock. I guess we should be using another one.

**Comment by rjeschmi** [comment]
Maybe we should create a feature branch of easyconfigs with some interesting examples merged?

**Comment by boegel** [comment]
The `*SOURCE_TAR_GZ` is standard YAML, see http://www.yaml.org/spec/1.2/spec.html#alias//
The definition of the alias with `&SOURCE_TAR_GZ` is injected by `eb` right before the `.yeb` file is parsed.

It's the best we found so far to as equivalent to the `SOURCE_TAR_GZ` constants we now have, since there's no discrimination between strings and variables in YAML otherwise.

**Comment by boegel** [comment]
A feature branch with easyconfigs makes sense to me. Question is how we should organise things there. Do we simply mix `.eb` and `.yeb` files in the same `<software>` directory?

**Comment by rjeschmi** [comment]
Yes, I think so about mixing. Though I think we expect to have only one .yeb per software (though we may allow inheritance of some sort)

File: docs/Writing_yeb_easyconfig_files.rst [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code::
+
+    %YAML 1.2
+    ---
+    easyblock: ConfigureMake
+
+    name: gzip
+    version: '1.6'  # FIXME bug: quotes are required here to make sure this is parsed a a string, not a floating point value
```
Comment (boegel) [inline docs/Writing_yeb_easyconfig_files.rst]: @Caylo: please look into this, make very sure somehow that `version` is string after parsing...

It's probably time we define the expected type for each of the easyconfig parameters...

Also, please include this `gzip.yeb` in the test easyconfigs, and include it in the tests for `.yeb`.

File: docs/Writing_yeb_easyconfig_files.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    sources: ['%(name)s-%(version)s.tar.gz']  # equivalent with [*SOURCE_TAR_GZ]
+
+See also :ref:`easyconfig_param_templates`.
+
+.. _easyconfig_yeb_format_syntax_dependencies:
+
+Dependencies
+~~~~~~~~~~~~
+
+*(WORK IN PROGRESS)*
```
Comment (boegel) [inline docs/Writing_yeb_easyconfig_files.rst]: @Caylo: this is all yours to figure out ;-)

File: docs/Writing_yeb_easyconfig_files.rst [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+As string concatenation is not supported in YAML, we defined a !join operator. It takes a list of values to concatenate
+as argument. This list can contain both hard strings and variables.
+Example::
+
+    !join [foo, bar]  # returns 'foobar'
+    vars:
+        - &f foo
+        - &b bar
+    !join [f, bar]  # returns 'foobar'
```
Comment (boegel) [inline docs/Writing_yeb_easyconfig_files.rst]: this works with `f`, no `*f` required?

File: docs/Writing_yeb_easyconfig_files.rst [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+Dependencies
+~~~~~~~~~~~~
+
+*(WORK IN PROGRESS)*
+Variables can be defined using standard YAML anchors (using a '&'). These are later referenced using an asterisk (*).
+Example::
+    vars:
+        - &compver 2015.3.187
+        - &gccsuff -GNU-4.9.3-2.25
```
Comment (boegel) [inline docs/Writing_yeb_easyconfig_files.rst]: I'll rework this a bit, this section should be on dependencies specifically, using internal variables is more generic, and should be described in a separate section.

**Comment by rjeschmi** [comment]
Add this feature to the list of experimental features in:

:ref:`experimental_features

File: docs/Writing_yeb_easyconfig_files.rst [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    using them currently requires enabling the use of experimental features (``--experimental``),
+    see also :ref:`experimental_features` .
+
+    An up-to-date overview of current progress on support for ``.yeb`` easyconfigs is available at
+    https://github.com/hpcugent/easybuild-framework/issues/1407.
+
+Useful links:
+
+* YAML syntax specification: http://www.yaml.org/spec/1.2/spec.html
+
```
Comment (rjeschmi) [inline docs/Writing_yeb_easyconfig_files.rst]: Is PyYAML a 1.2 parser? They advertise as 1.1 (though I imagine there are some extensions)

Comment (boegel) [inline docs/Writing_yeb_easyconfig_files.rst]: Hmm, good point.

Although, PyYAML is not complaining when we're using `%YAML 1.2` as a header.

**Comment by rjeschmi** [comment]
Merge it as is, things look fine and generally good. We'll work with it and update as we learn more.

**Comment by boegel** [comment]
@rjeschmi: makes sense, let's do that indeed, thanks for the review!

**Pull request merged**